design/dpop.md

```diff
@@ -0,0 +1,259 @@
+# OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)
+
+* **Status**: Notes
+* **JIRA**: TBD
+
+## Motivation
+
+DPoP is a mechanism to prevent illegal API calls from succeeding only with a stolen access token. This is done via augmenting the calls to OAuth 2.0 endpoints as well as resource servers with a special DPoP header that proves possession of a private key, and binding the issued tokens to that key. Please refer to the Resources section for more info on DPoP.    
+
+## Implementation Details
+
+DPoP is a cross-cutting concept that would affect multiple Keycloak components, namely core libraries, OAuth/OIDC endpoints, adapters, models and admin UI.
+
+It is suggested that DPoP should be set up on a per-client basis, with the three options available that should be interpreted as follows:
+
+|**Value** | **Keycloak Server** | **Keycloak Adaptors** |
+| --- | --- | --- |
+| Required | OAuth endpoints (where applicable) should require a DPoP header and a DPoP-bound token; should return an error otherwise. | Protected resources should require a DPoP header and a DPoP-bound token, or should return an error otherwise. |
+| Optional | _Token Endpoint_: if DPoP header is present, the returned token should be DPoP-bound and should contain a `cnf` claim with a JWK thumbprint; should be a regular token otherwise  <br/><br/>_UserInfo endpoint_: if the access token is DPoP-bound, a matching DPoP header should be required; no header should be required otherwise. | Protected resources should require a matching DPoP header if the supplied access token is DPoP-bound; should ignore the header otherwise. |
+| Disabled | DPoP headers should be ignored; issued tokens shouldn’t contain JWK thumbprints. | DPoP headers and token JWK thumbprints should be ignored. |
+
```

Comment:
By looking at [DPoP's draft](https://tools.ietf.org/id/draft-ietf-oauth-dpop-02.html#name-dpop-access-token-request), if we make refresh token DPoP-bound, we need to use the same public key whenever refreshing tokens. Therefore, we need to use and keep the same public key until the first issued original refresh token expires or is revoked. Considering that the lifetime of the refresh token is much longer than the access token, a SPA as the client without its secure storage needs to keep it longer, which might be not good from security perspective.

Reply:
Good point, will add a section on refresh tokens. The above should be relevant to offline tokens as well, since those are basically "immortal" refresh tokens. Hence, the keypair should be persisted for the lifetime of the offline token.

Also, AFAIK Keycloak always issues a new refresh token upon refresh ([link](https://github.com/keycloak/keycloak/blame/master/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L343)), so this will result in a chain of refresh tokens bound to the same DPoP keypair; so the keypair should be kept for the duration of the entire session.

Reply:
Yes, I think so.
It might be better that we have a chance to change the key when we refresh the token. However, I have not found how to do it in [the DPoP specification](https://tools.ietf.org/id/draft-ietf-oauth-dpop-02.html#section-5). 

This DPoP specification treats the following things individually.
- issue holder-of-key bound token (by AS)
- verify holder-of-key bound token (by RS)

and does not treat them simultaneously (eg. refresh tokens)
- verify holder-of-key bound old token and issue holder-of-key bound new token (by AS)
- 
which means that we only use the same key for both old and new token. Also it leads to using the same key during the refresh token (also offline token in OIDC) is active. Refreshing keys is good from security perspective but we cannot do it.
 
It might be better to have the way of using separate key for each old and new token and moreover be specified by DPoP specification.

Reply:
Excellent point - the DPoP spec currently doesn't mention any means for key rotation; I think we should approach the authors of the DPoP spec with this. The solution might be as simple as providing two DPoP proofs in a single header (for the special case of token refresh):

```
POST /token HTTP/1.1
Host: server.example.com
Content-Type: application/x-www-form-urlencoded;charset=UTF-8
DPoP: eyJ0eXAiO... eyJ0eXAiO...

grant_type=refresh_token
&refresh_token=Q..Zkm29lexi8VnWg2zPW1x-tgGad0Ibc3s3EwM_Ni4-g
```

This would allow the server to use the first DPoP proof for verification, and the second one for the issuance of a new token set. Or perhaps those could be two different headers.

Reply:
According to https://tools.ietf.org/html/draft-ietf-oauth-dpop-03#section-5 , 

> When an authorization server supporting DPoP issues a refresh token to a public client that presents a valid DPoP proof at the token endpoint, the refresh token MUST be bound to the respective public key. 
> The binding MUST be validated when the refresh token is later presented to get new access tokens.  As a result, such a client MUST present a DPoP proof for the same key that was used to obtain the refresh token each time that refresh token is used to obtain a new access token.  

Therefore, by following this paragraph, we need to use the same key that was used to get the first access and refresh token whenever we refresh tokens. As a result, all access and refresh tokens are bound with the same key until refreshing tokens is invalidated (ex. expiration of the first refresh token, revoking it).

Reply:
Yes, this missing support for key rotation looks like a limitation of DPoP specification...

One possible workaround for this is to make sure that client does full redirect for the SSO re-authentication instead of refresh tokens. For example we can document for the javascript adapter (and maybe other parts of the documentation) that client is encouraged to do full SSO redirect after some time (EG. every 30 minutes) or after some count of refreshes (EG. 5). SSO redirect will use same userSession and hence there is not additional data needed on the server side.

This may not be ideal for some kind of applications as SSO redirect can cause the lose of some state of the javascript application in the browser.  And DPoP with refresh token is intended to be used primarily for the public clients according to the specification and hence primarily javascript applications. But at least we can document that such possibility exists as it can be used at least by some applications. WDYT?

Reply:
@mposolda agree on that; we can mention this as "best practices" in the JS adapter docs.

Just in case, here is the stance on DPoP key rotation from from the spec co-author:

> (...) the general rough consensus thinking has been
that allowing for DPoP key rotation for refresh tokens and public clients
(the only case where it's relevant) didn't add enough value to justify the
added complexity. It doesn't help with the threat model for in-browser
applications. And mobile applications have really good options for key
storage - to the point that the kind of event that might compromise a DPoP
key would involve a lot more than key rotation to cleanup from.

[(ML thread)](https://mailarchive.ietf.org/arch/msg/oauth/65Aip95ixdP8PplzFXuIfREERvc/)

Reply:
@dteleguin Thanks for the update and referencing this discussion!